File: server/routes/versions.ts
import { Hono } from 'hono'
import { v4 as uuidv4 } from 'uuid'
import { z } from 'zod'
import { db } from '../db/sqlite'
import { requireAuth, getRequestUser } from '../middleware/auth'

const versions = new Hono()

const CreateVersionSchema = z.object({
  project_id: z.string().uuid('项目 ID 无效'),
  content: z.string(),
  change_summary: z.string().optional(),
})

const UpdateVersionSchema = z.object({
  content: z.string().min(1, '版本内容不能为空'),
})

versions.use('*', requireAuth)

function userOwnsProject(projectId: string, userId: string): boolean {
  const project = db
    .prepare("SELECT id FROM projects WHERE id = ? AND user_id = ? AND status != 'deleted'")
    .get(projectId, userId)
  return Boolean(project)
}

versions.get('/', (c) => {
  const user = getRequestUser(c)
  const projectId = c.req.query('project_id')

  if (!projectId) return c.json({ error: '缺少项目 ID' }, 400)
  if (!userOwnsProject(projectId, user.id)) {
    return c.json({ error: '项目不存在或无权访问' }, 404)
  }

  const results = db
    .prepare(
      `SELECT id, project_id, created_by, change_summary, timestamp
       FROM versions
       WHERE project_id = ?
       ORDER BY timestamp DESC`,
    )
    .all(projectId)

  return c.json(results, 200)
})

versions.post('/', async (c) => {
  try {
    const user = getRequestUser(c)
    const body = await c.req.json()
    const parsed = CreateVersionSchema.safeParse(body)

    if (!parsed.success) {
      return c.json({ error: parsed.error.issues[0].message }, 400)
    }

    const { project_id, content, change_summary } = parsed.data
    if (!userOwnsProject(project_id, user.id)) {
      return c.json({ error: '项目不存在或无权访问' }, 404)
    }

    const versionId = uuidv4()

    db.transaction(() => {
      db.prepare(
        `INSERT INTO versions (id, project_id, created_by, content, change_summary, timestamp)
         VALUES (?, ?, ?, ?, ?, CURRENT_TIMESTAMP)`,
      ).run(versionId, project_id, user.id, content, change_summary || '')

      db.prepare('UPDATE projects SET updated_at = CURRENT_TIMESTAMP WHERE id = ?').run(project_id)
    })()

    return c.json({ id: versionId, project_id, timestamp: new Date().toISOString() }, 201)
  } catch (err) {
    console.error('Create version error', err)
    return c.json({ error: '服务器内部错误' }, 500)
  }
})

versions.get('/detail', (c) => {
  const user = getRequestUser(c)
  const id = c.req.query('id')

  if (!id) return c.json({ error: '缺少版本 ID' }, 400)

  const version = db
    .prepare(
      `SELECT v.*
       FROM versions v
       JOIN projects p ON v.project_id = p.id
       WHERE v.id = ? AND p.user_id = ? AND p.status != 'deleted'`,
    )
    .get(id, user.id)

  if (!version) return c.json({ error: '版本不存在或无权访问' }, 404)

  return c.json(version, 200)
})

versions.put('/detail', async (c) => {
  try {
    const user = getRequestUser(c)
    const id = c.req.query('id')
    if (!id) return c.json({ error: '缺少版本 ID' }, 400)

    const body = await c.req.json()
    const parsed = UpdateVersionSchema.safeParse(body)

    if (!parsed.success) {
      return c.json({ error: parsed.error.issues[0].message }, 400)
    }

    const version = db
      .prepare(
        `SELECT v.id, v.project_id
         FROM versions v
         JOIN projects p ON v.project_id = p.id
         WHERE v.id = ? AND p.user_id = ? AND p.status != 'deleted'`,
      )
      .get(id, user.id) as { id: string; project_id: string } | undefined

    if (!version) return c.json({ error: '版本不存在或无权访问' }, 404)

    db.transaction(() => {
      db.prepare('UPDATE versions SET content = ?, timestamp = CURRENT_TIMESTAMP WHERE id = ?').run(
        parsed.data.content,
        id,
      )
      db.prepare('UPDATE projects SET updated_at = CURRENT_TIMESTAMP WHERE id = ?').run(version.project_id)
    })()

    return c.json({ success: true }, 200)
  } catch (err) {
    console.error('Update version error', err)
    return c.json({ error: '服务器内部错误' }, 500)
  }
})

versions.delete('/detail', (c) => {
  const user = getRequestUser(c)
  const id = c.req.query('id')
  if (!id) return c.json({ error: '缺少版本 ID' }, 400)

  const result = db
    .prepare(
      `DELETE FROM versions
       WHERE id = ?
         AND project_id IN (SELECT id FROM projects WHERE user_id = ? AND status != 'deleted')`,
    )
    .run(id, user.id)

  if (result.changes === 0) return c.json({ error: '版本不存在或无权访问' }, 404)

  return c.json({ success: true }, 200)
})

export default versions

File: backend/internal/db/versions.go
package db

import (
	"database/sql"

	"github.com/google/uuid"
)

// CreateVersion 为项目插入一条版本，返回版本 id。
func (s *Store) CreateVersion(projectID, createdBy, content, changeSummary string) (string, error) {
	id := uuid.NewString()
	_, err := s.db.Exec(
		`INSERT INTO versions (id, project_id, created_by, content, change_summary, timestamp)
		 VALUES (?, ?, NULLIF(?, ''), ?, ?, CURRENT_TIMESTAMP)`,
		id, projectID, createdBy, content, changeSummary,
	)
	if err != nil {
		return "", err
	}
	return id, nil
}

// ListVersions 返回项目全部版本元信息，按时间倒序。
func (s *Store) ListVersions(projectID string) ([]Version, error) {
	rows, err := s.db.Query(
		`SELECT id, project_id, created_by, change_summary, timestamp
		 FROM versions WHERE project_id = ? ORDER BY timestamp DESC, rowid DESC`,
		projectID,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []Version{}
	for rows.Next() {
		var v Version
		if err := rows.Scan(&v.ID, &v.ProjectID, &v.CreatedBy, &v.ChangeSummary, &v.Timestamp); err != nil {
			return nil, err
		}
		out = append(out, v)
	}
	return out, rows.Err()
}

// GetVersion 返回单个版本详情（含 content）。
func (s *Store) GetVersion(id string) (*VersionDetail, error) {
	var v VersionDetail
	err := s.db.QueryRow(
		`SELECT id, project_id, created_by, content, change_summary, timestamp
		 FROM versions WHERE id = ?`,
		id,
	).Scan(&v.ID, &v.ProjectID, &v.CreatedBy, &v.Content, &v.ChangeSummary, &v.Timestamp)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &v, nil
}

// GetVersionOwnedByUser 联表校验版本属于用户本人项目后返回。
func (s *Store) GetVersionOwnedByUser(id, userID string) (*VersionDetail, error) {
	var v VersionDetail
	err := s.db.QueryRow(
		`SELECT v.id, v.project_id, v.created_by, v.content, v.change_summary, v.timestamp
		 FROM versions v JOIN projects p ON v.project_id = p.id
		 WHERE v.id = ? AND p.user_id = ? AND p.status != 'deleted'`,
		id, userID,
	).Scan(&v.ID, &v.ProjectID, &v.CreatedBy, &v.Content, &v.ChangeSummary, &v.Timestamp)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &v, nil
}

// UpdateVersionContent 更新版本内容。
func (s *Store) UpdateVersionContent(id, content string) (bool, error) {
	res, err := s.db.Exec("UPDATE versions SET content=? WHERE id=?", content, id)
	if err != nil {
		return false, err
	}
	n, _ := res.RowsAffected()
	return n > 0, nil
}

// DeleteVersion 删除版本。
func (s *Store) DeleteVersion(id string) (bool, error) {
	res, err := s.db.Exec("DELETE FROM versions WHERE id=?", id)
	if err != nil {
		return false, err
	}
	n, _ := res.RowsAffected()
	return n > 0, nil
}

// VersionCount 统计全部版本数。
func (s *Store) VersionCount() (int, error) {
	var n int
	err := s.db.QueryRow("SELECT COUNT(*) FROM versions").Scan(&n)
	return n, err
}

// LatestVersionOfProject 返回某项目最新一版（按时间倒序），无版本时返回 nil。
func (s *Store) LatestVersionOfProject(projectID string) (*VersionDetail, error) {
	var v VersionDetail
	err := s.db.QueryRow(
		`SELECT id, project_id, created_by, content, change_summary, timestamp
		 FROM versions WHERE project_id = ? ORDER BY timestamp DESC, rowid DESC LIMIT 1`,
		projectID,
	).Scan(&v.ID, &v.ProjectID, &v.CreatedBy, &v.Content, &v.ChangeSummary, &v.Timestamp)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &v, nil
}